Ansible module to enable the user to provision CPGs on HPE StoreServ 3PAR device

farhan7500 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Ansible modules which enable the user to provision CPGs on HPE StoreServ 3PAR device using Python 3PAR client to communicate over WSAPI REST interface. Ansible behaves as the orchestrator tool.
<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
storage/hpe3par/hpe3par_cpg

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (review-address2 661fbedf3a) last updated 2018/06/01 05:05:09 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/farhan/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/farhan/workspace/ansible/lib/ansible
  executable location = /home/farhan/workspace/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

Commit:
Added storage modules and unit tests

Removed unnecessary file

Fixing pep8 issues

Adding reusable documentation to the data fragment file

Fixing issues reported by module validation in documentation

Fixed issues: 10549, 10553

Fixed unit tests

Added future imports and removed license from init py file

Resolved import issue reported by ansible sanity test

Fixed unit tests

Adding CPG management module

Fixing sanity test issues

Resolving line ending issues

Resolved import issue reported by ansible sanity test

Correcting the error message when resource is more than 31 char

Update hpe3par_cpg.py

Fix for defect 10556: Unable to modify volume. Persona not set

Fix for defect 10556: Unable to modify volume. Persona not set

Fixing pep8 changes

Fixed pep8 issues

Added more pep8 related fixes

Added more pep8 related fixes

Fixed rebase issue in cps test

Adding GPL3.0+ License (#7)

* Update hpe3par.py

* Update hpe3par_cpg.py

* Update hpe3par_cpg.py

Adding GPLV3+ only license

Fix anity test

Addressing review comments

Fixing undefined variable error

Addressed review comments and added more unit tests (#8)

Added maintainers

Fixed sanity test

Update test_hpe3par_cpg.py

Renamed the module as per the review comments (#9)

* Added maintainers

* Renaming the module to remove the vendor name from the module

* remove pep8 related issues

Modify the requirements to add version of 3par sdk

Review address (#10)

* Unit is no longer a different paramter in the task

* Fixing unit tests

* Taking into account the default values

Fixing issue with default values

fixing unit test

Adding more flexibility and checks to the size and it's units

Fixed pylint extra line issue

Fixing conflicts in BOTMETA.yml

Made changes according to the review comments

Updating ansible version

Refactored the code to move reusable method to module utils

Modified the GPLv3+ license text according to the review comments. Modified the module utils license to BSD

Corrected the BOTMETA entry. Fixed the pep8 issues

Fixing entries in BOTMETA.yml

Removed ansible version, renamed the version to address the name review comments

Reverting name change for module to revert test failures

Set size and raid type and now reuqired together based on the review comment

Fixed unit tests

Removing default value as raid and set size are now required together

Renaming the modules to ss_3par_*

Resolving the pep8 issues

Added storage modules and unit tests

Removed unnecessary file

Fixing pep8 issues

Adding reusable documentation to the data fragment file

Fixing issues reported by module validation in documentation

Fixed issues: 10549, 10553

Fixed unit tests

Added future imports and removed license from init py file

Resolved import issue reported by ansible sanity test

Fixed unit tests

Adding CPG management module

Fixing sanity test issues

Resolving line ending issues

Resolved import issue reported by ansible sanity test

Correcting the error message when resource is more than 31 char

Update hpe3par_cpg.py

Fix for defect 10556: Unable to modify volume. Persona not set

Fix for defect 10556: Unable to modify volume. Persona not set

Fixing pep8 changes

Fixed pep8 issues

Added more pep8 related fixes

Added more pep8 related fixes

Fixed rebase issue in cps test

Adding GPL3.0+ License (#7)

* Update hpe3par.py

* Update hpe3par_cpg.py

* Update hpe3par_cpg.py

Adding GPLV3+ only license

Fix anity test

Addressing review comments

Fixing undefined variable error

Addressed review comments and added more unit tests (#8)

Added maintainers

Fixed sanity test

Update test_hpe3par_cpg.py

Renamed the module as per the review comments (#9)

* Added maintainers

* Renaming the module to remove the vendor name from the module

* remove pep8 related issues

Modify the requirements to add version of 3par sdk

Review address (#10)

* Unit is no longer a different paramter in the task

* Fixing unit tests

* Taking into account the default values

Fixing issue with default values

fixing unit test

Adding more flexibility and checks to the size and it's units

Fixed pylint extra line issue

Fixing conflicts in BOTMETA.yml

Made changes according to the review comments

Updating ansible version

Refactored the code to move reusable method to module utils

Modified the GPLv3+ license text according to the review comments. Modified the module utils license to BSD

Corrected the BOTMETA entry. Fixed the pep8 issues

Removed ansible version, renamed the version to address the name review comments

Reverting name change for module to revert test failures

Set size and raid type and now reuqired together based on the review comment

Fixed unit tests

Removing default value as raid and set size are now required together

Renaming the modules to ss_3par_*

Resolving the pep8 issues

Update lib/ansible/modules/storage/hpe3par/ss_3par_cpg.py

Co-Authored-By: farhan7500 <farhan7500@gmail.com>

Update lib/ansible/modules/storage/hpe3par/ss_3par_cpg.py

Co-Authored-By: farhan7500 <farhan7500@gmail.com>

Fixed rebase issue

Fixed rebase issuein BOTMETA.yml

mattclay commented:
The unrelated errors have been fixed. I've restarted CI for this PR.

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+
+DOCUMENTATION = r'''
+---
+author: "Hewlett Packard Enterprise (ecostor@groups.ext.hpe.com)"
```
it should be a [GitHub ID](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block) instead of an email.

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+---
+author: "Hewlett Packard Enterprise (ecostor@groups.ext.hpe.com)"
+description:
+  - "Create and delete CPG on HPE 3PAR."
```
A link to CPG or 3PAR could be added using [formatting function](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#formatting-functions).

pilou- replied:
Not done but not a mandatory change. Just in case, here is an [example](https://github.com/ansible/ansible/blob/f89d873698c0d22c2f1692938854db0fe3d0acb0/lib/ansible/modules/cloud/amazon/aws_kms_facts.py#L170).

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+      - MAG
+    description:
+      - "Specifies that the layout must support the failure of one port pair,
+       one cage, or one magazine.\n"
```
Line feed could be removed (same for `growth_warning`)

farhan7500 replied:
Line feed removed

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+        storage_system_username="{{ storage_system_username }}"
+        storage_system_password="{{ storage_system_password }}"
+        state=present
+        cpg_name="{{ cpg_name }}"
```
Values could be used instead of templates (for growth_* parameters and parameters below).

farhan7500 replied:
Removed templates

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+        size_mib = size * 1048576
+    elif size_unit == 'MiB':
+        size_mib = size
+    return int(size_mib)
```
`convert_to_binary_multiple` is used with `growth_limit`, `growth_increment` and `growth_warning` which are all `float` (due to `argument_spec`):
- should not `growth_limit`, `growth_increment` and `growth_warning` be `int` ?
- if not, `test_convert_to_binary_multiple` should test this method using float parameter

farhan7500 replied:
3PAR API only accepts the size in MiB. However, we allow the user to provide the size in GiB and TiB as well which can be floats. However, these user provided sizes are converted into MiB for API consumption which cannot be a float. Hence casting the final value into int

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+author: "Hewlett Packard Enterprise (ecostor@groups.ext.hpe.com)"
+description:
+  - "Create and delete CPG on HPE 3PAR."
+module: hpe3par_cpg
```
I would consider calling this simply **3par_cpg** and create a directory **3par**. We tend not to create technology-named modules/directories rather than vendor-named things. (We have historical exceptions)

For instance, the Cisco ACI modules are named: network/aci/aci_tenant.py and not network/cisco/cisco_aci_tenant.py

farhan7500 replied:
We don't want the file to begin with a number, hence keeping it prefixed with hpe3par. Changed the directory name to hpe3par to be more specific.

farhan7500 replied:
I have changed the name of the module to cpg_3par, subsequent modules will also be names xyz_3par

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+
+DOCUMENTATION = r'''
+---
+author: "Hewlett Packard Enterprise (ecostor@groups.ext.hpe.com)"
```
This needs to be using Github handles, so something like:
```yaml
author:
- Your Name (@farhan7500)
```

This way changes to this file can be merged by user @farhan7500 through ansibot.

farhan7500 replied:
Done

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+---
+author: "Hewlett Packard Enterprise (ecostor@groups.ext.hpe.com)"
+description:
+  - "Create and delete CPG on HPE 3PAR."
```
No need to quote strings, unless in specific cases.

farhan7500 replied:
Removed the quotes

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+options:
+  cpg_name:
+    description:
+      - "Name of the CPG."
```
Again, no need to quote all strings.

farhan7500 replied:
Removed the quotes

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+      - SSD
+    description:
+      - "Specifies that physical disks must have the specified device type."
+    required: false
```
If parameters are not required, you don't need to add this.
It helps skim docs if only the required parameters have the "required" key.

farhan7500 replied:
Removed the required parameter where its false

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+    default: -1.0
+    description:
+      - "Specifies that the threshold of used logical disk space when exceeded
+       results in a warning alert.\n"
```
That newline is probably a conversion error ?

farhan7500 replied:
Removed newline

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+      - "Whether the specified CPG should exist or not."
+    required: true
+extends_documentation_fragment: hpe3par
+short_description: "Manage HPE 3PAR CPG"
```
I'd put the short_description at the top, as we used to for other modules.

farhan7500 replied:
Moved the short description to the top

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+EXAMPLES = r'''
+    - name: Create CPG "{{ cpg_name }}"
+      hpe3par_cpg:
+        storage_system_ip="{{ storage_system_ip }}"
```
It is a lot easier for users if examples show real-life content, rather than stub templates.
Now the example does not really add anything.

Also, it should be using the YAML syntax, and not the old key=value syntax.

dagwieers replied:
Default connection parameters are called "host", "username", "password", I think it helps if your module is consistent with the standard parameter names. If you still like, you can add aliases, but I don't see the value.

farhan7500 replied:
Modified the examples to show real life contents as mentioned.

In case of 3PAR, host has a different meaning where the volumes are presented. Therefore, keeping it as is to keep it unambiguous

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+
+from ansible.module_utils.basic import AnsibleModule
+try:
+    from hpe3par_sdk import client
```
We tend to add something like HAS_3PAR = True and HAS_3PAR = False.
See other modules doing this.

farhan7500 replied:
Done

dagwieers commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+    client = None
+
+
+def convert_to_binary_multiple(size, size_unit):
```
We have this functionality already in Ansible.
Personally I would make the interface support value + unit, rather than 2 parameters.

farhan7500 replied:
Modified to reuse the functionality from module_utils

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+        set_size,
+        high_availability,
+        disk_type):
+    if storage_system_username is None or storage_system_password is None:
```
You might want to add a `module` parameter and use `module.fail_json` for these 4 checks. Another way to do might be to perform these checks before calling `create_cpg`.

farhan7500 replied:
Since these are required parameters, there was no need for the null checks. Hence removed.

For all other none checks where the parameter is only required for a certain action, modified the code and using module.fail_json

pilou- replied:
In fact, end user is able to explicitly use a `None` value, even for required parameter. The checks have been moved in [`main`](https://github.com/ansible/ansible/pull/39656/files#diff-3c0fdc61c06620158bc670b5868dd735R277), that's ok.

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+    high_availability = module.params["high_availability"]
+    disk_type = module.params["disk_type"]
+
+    wsapi_url = 'https://%s:8080/api/v1' % storage_system_ip
```
Should not the port be a parameter too (something like `storage_system_ip`) ?

farhan7500 replied:
In case of 3PAR port will always be 8080

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+    disk_type = module.params["disk_type"]
+
+    wsapi_url = 'https://%s:8080/api/v1' % storage_system_ip
+    client_obj = client.HPE3ParClient(wsapi_url)
```
`secure=True` must be used with `client.HPE3ParClient` or it can be another parameter.

farhan7500 replied:
Added secure=True

pilou- replied:
`secure` has been added. I would rather use `True` as a default value:
https://github.com/ansible/ansible/blob/4e6a50cf4433ec96edc0ee7ac6d20eb0c392272c/lib/ansible/module_utils/hpe3par.py#L31-L34

pilou- commented on lib/ansible/utils/module_docs_fragments/hpe3par.py:
```diff
@@ -0,0 +1,42 @@
+      required: true
+
+requirements:
+  - "3PAR OS - 3.2.2 MU6, 3.3.1 MU1"
```
Should not the `3PAR OS - 3.2.2 MU6, 3.3.1 MU` requirement checked somewhere ?

`hpe3par_sdk` doesn't need to be checked since there isn't any previous release.

farhan7500 replied:
The 3PAR firmware version check is done at the library level

pilou- replied:
Because:
> Logout has been made idempotent in the hpe3par_sdk library now. It will not raise an exception is not already logged in as well ([here](https://github.com/ansible/ansible/pull/39656#discussion_r190213227))

and
https://github.com/ansible/ansible/blob/4e6a50cf4433ec96edc0ee7ac6d20eb0c392272c/lib/ansible/utils/module_docs_fragments/hpe3par.py#L39
then the `hpe3par_sdk` version should be checked (preferably in `module_utils.hpe3par` in order to share the check between modules).

pilou- commented on test/units/modules/storage/hpe3par/test_hpe3par_cpg.py:
```diff
@@ -0,0 +1,343 @@
+
+class TestHpe3parCPG(unittest.TestCase):
+
+    fields = {
```
`fields` in `hpe3par_cpg.py` could be a module attribute and then reused there.

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+            "choices": ['present', 'absent'],
+            "type": 'str'
+        },
+        "storage_system_ip": {
```
Often parameters documented in a doc fragment are defined in [`lib/ansible/module_utils`](https://github.com/ansible/ansible/tree/devel/lib/ansible/module_utils/) (doc fragments allow to share doc between modules, `lib/ansible/module_utils` allows to share code)

farhan7500 replied:
Done

pilou- replied:
You could have kept parameters which are not defined in the doc fragment in this file (except if you plan do reuse `cpg_argument_spec` in another module.

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+            client_obj.deleteCPG(cpg_name)
+        else:
+            return (True, False, "CPG does not exist", {})
+    except Exception as e:
```
`hpe3parclient.exceptions.ClientException` could be used instead of `Exception`.

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+        storage_system_username,
+        storage_system_password,
+        cpg_name):
+    if storage_system_username is None or storage_system_password is None:
```
These 3 checks are copied from `create_cpg` method: these checks need to be done before calling `create_cpg`/`delete_cpg` (in `main` or in a dedicated method).
In the same way, `client_obj.login` could be called before calling `create_cpg`/`delete_cpg`.

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+    except Exception as e:
+        return (False, False, "CPG delete failed | %s" % e, {})
+    finally:
+        client_obj.logout()
```
No need to call logout when `login` raises `hp3parclient.exceptions.HTTPUnauthorized` exception (otherwise it seems `logout` will raise another exception).

farhan7500 replied:
Logout has been made idempotent in the hpe3par_sdk library now. It will not raise an exception is not already logged in as well

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+
+    # States
+    if module.params["state"] == "present":
+        return_status, changed, msg, issue_attr_dict = create_cpg(
```
`issue_attr_dict` is always an empty dict, remove it.

farhan7500 replied:
Removed

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+            disk_type
+        )
+    elif module.params["state"] == "absent":
+        return_status, changed, msg, issue_attr_dict = delete_cpg(
```
`issue_attr_dict` is always an empty dict, remove it.

farhan7500 replied:
Removed

pilou- commented on test/units/modules/storage/hpe3par/test_hpe3par_cpg.py:
```diff
@@ -0,0 +1,343 @@
+
+
+import mock
+import unittest
```
Note unit tests of new modules must [use `pytest`](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#unit-testing) instead of `unittest`. Don't hesitate to ask if you need guidance for migrate to pytest.

farhan7500 replied:
Using pytest now

pilou- commented on lib/ansible/modules/storage/hpe/hpe3par_cpg.py:
```diff
@@ -0,0 +1,435 @@
+    if not validate_set_size(raid_type, set_size):
+        return (False, False, "Set size not part of RAID set", {})
+    try:
+        validate_set_size(raid_type, set_size)
```
This 2nd call to `validate_set_size` must be removed.

farhan7500 replied:
Removed

pilou- reviewed (commented):
Some changes are required. Unit tests are provided, that's great !

farhan7500 commented:
@dagwieers @pilou- 
We have addressed the review comments, kindly review.

Thanks.

farhan7500 commented:
ready_for_review

cc @pilou- @dagwieers - Thanks for your review. We have addressed the review comments and the PR is ready to be merged. Please have a look. Thanks!

farhan7500 commented:
ready_for_review

farhan7500 commented:
Closing and opening PR to rerun CI tests

farhan7500 commented:
Reopening PR to rerun CI tests

farhan7500 commented:
ready_for_review

gautamphegde commented:
bot_status

gautamphegde commented:
bot_broken

jctanner commented:
@farhan7500 please squash your commits. >50 commits causes the bot to mark the PR as "bad" and therefore ignored.

farhan7500 commented:
bot_status

farhan7500 commented:
ready_for_review

dagwieers commented:
I made [a comment](https://github.com/ansible/ansible/pull/39656#pullrequestreview-119193759) about the name, we usually name modules after the product only, not the vendor. So my suggestion was to use **3par_cpg** instead. The author does not want the module to start with a number.

dagwieers commented:
As a follow-up, we named the HP iLO modules hpilo, but I guess we now have to call it [hpeilo](https://www.hpe.com/be/en/servers/integrated-lights-out-ilo.html) ? ;-)

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+
+
+def validate_set_size(raid_type, set_size):
+    if raid_type is not None or set_size is not None:
```
Should not that be `if raid_type:` ? Because:
- Default value for `set_size` is `-1` (in both `argument_spec` and documentation)
- code below works without testing `set_size`and handle both `None` and `-1` value

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+        if not client_obj.cpgExists(cpg_name):
+            ld_layout = dict()
+            disk_patterns = []
+            if disk_type is not None and disk_type:
```
`if disk_type:` could be used here (`bool(None) is False`). Note that the same pattern (`if x is not None and x:`) is reused several times.

pilou- commented on lib/ansible/module_utils/hpe3par.py:
```diff
@@ -0,0 +1,102 @@
+        "raid_type": {
+            "required": False,
+            "type": "str",
+            "choices": ['R0', 'R1', 'R5', 'R6']
```
Should not that be [`hpe3par_sdk.client.HPE3ParClient.keys()`](https://github.com/HewlettPackard/hpe3par_python_sdk/blob/v1.0.1/hpe3par_sdk/client.py#L48) instead of list of hardcoded values ?

farhan7500 replied:
I am creating the ansible module in main,

module = AnsibleModule(argument_spec=hpe3par.cpg_argument_spec())
and then failing using module.fail_json method if the library is not installed.

If I replace the choices by referring the client's RAID_MAP keys and if the library is missing, it will throw an exception even before the library check. At this point, I don't have the module to forcefully fail (fail_json - with the appropriate message). Is there another way of checking for the existence of library before the ansible module is created?

pilou- replied:
Indeed, there isn't an easy way to do that.

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+        disk_type):
+    try:
+        if not validate_set_size(raid_type, set_size):
+            return (False, False, "Set size not part of RAID set", {})
```
Error message could mention both `raid_type` and `set_size` values.

farhan7500 replied:
Added the values of raid_type and set_size

pilou- commented on test/units/modules/storage/hpe3par/test_hpe3par_cpg.py:
```diff
@@ -0,0 +1,217 @@
+    hpe3par_cpg.validate_set_size = mock.Mock(return_value=True)
+    hpe3par_cpg.cpg_ldlayout_map = mock.Mock(
+        return_value={'RAIDType': 4, 'HA': 1})
+    hpe3par_cpg.convert_to_binary_multiple = mock.Mock(return_value=1000)
```
I don't see any reason to mock the following methods:
- `validate_set_size`
- `cpg_ldlayout_map`
- `convert_to_binary_multiple` method.

These methods don't depend on any external resources (they don't perform any API request).

farhan7500 replied:
Removed mocking for convert_to_binary_multiple but validate_set_size and cpg_ldlayout_map depend on the library. Eg,

def cpg_ldlayout_map(ldlayout_dict):
    if ldlayout_dict['RAIDType'] is not None and ldlayout_dict['RAIDType']:
        ldlayout_dict['RAIDType'] = **client.HPE3ParClient.RAID_MAP**[
            ldlayout_dict['RAIDType']]['raid_value']
    if ldlayout_dict['HA'] is not None and ldlayout_dict['HA']:
        ldlayout_dict['HA'] = getattr(
            client.HPE3ParClient, ldlayout_dict['HA'])
    return ldlayout_dict

and,

def validate_set_size(raid_type, set_size):
    if raid_type:
        set_size_array = **client.HPE3ParClient.RAID_MAP[raid_type]['set_sizes']**
        if set_size in set_size_array:
            return True
    return False

Also, these methods are being tested individually as well with mocked library interface

pilou- replied:
I missed the fact that the whole python module is mocked ! I would have limited mocks to requests to external resources but that's ok.

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+                'HA': high_availability,
+                'diskPatterns': disk_patterns}
+            ld_layout = cpg_ldlayout_map(ld_layout)
+            if growth_increment is not None:
```
Default value of `growth_increment`, `growth_limit` and `growth_warning` is `-1.0`. A `ValueError` will be raised by `convert_to_binary_multiple` each time the default value is used for any of these parameters:
```
In [1]: from ansible.modules.storage.hpe3par.hpe3par_cpg import convert_to_binary_multiple

In [2]: convert_to_binary_multiple(-1.0, 'GiB')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-0a8ba09b9258> in <module>()
----> 1 convert_to_binary_multiple(-1.0, 'GiB')

ansible/lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py in convert_to_binary_multiple(size, size_unit)
    180         suffix = 'T'
    181 
--> 182     size_kib = basic.human_to_bytes(str(size) + suffix)
    183     return int(size_kib / (1024 * 1024))
    184 

ansible/lib/ansible/module_utils/basic.py in human_to_bytes(number, default_unit, isbits)
    634         num = float(m.group(1))
    635     except:
--> 636         raise ValueError("human_to_bytes() can't interpret following number: %s (original input string: %s)" % (m.group(1), number))
    637 
    638     unit = m.group(2)

ValueError: human_to_bytes() can't interpret following number:  (original input string: -1.0G)
```

farhan7500 replied:
Made the required changes in convert_to_binary_multiple

pilou- commented on test/units/modules/storage/hpe3par/test_hpe3par_cpg.py:
```diff
@@ -0,0 +1,217 @@
+
+
+def test_convert_to_binary_multiple():
+    assert hpe3par_cpg.convert_to_binary_multiple(1, 'MiB') == 1
```
`convert_to_binary_multiple` is always called with a float: first parameter must be a float, not an integer (as pointed out in this [previous review](https://github.com/ansible/ansible/pull/39656#discussion_r187408218)).

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+        raid_type,
+        set_size,
+        high_availability,
+        disk_type):
```
A previous version checked values of `storage_system_username` and `storage_system_password`. These checks has been removed. You [wrote](https://github.com/ansible/ansible/pull/39656#discussion_r190212665):
> Since these are required parameters, there was no need for the null checks. Hence removed.

This isn't true: use could provide explicitly a `null`/`None` value. `required=True` doesn't imply `value will not be None`.

farhan7500 replied:
Added the checks for username and password

pilou- commented on lib/ansible/utils/module_docs_fragments/hpe3par.py:
```diff
@@ -0,0 +1,42 @@
+      required: true
+
+requirements:
+  - 3PAR OS - 3.2.2 MU6, 3.3.1 MU1
```
You [wrote](https://github.com/ansible/ansible/pull/39656#discussion_r190212913) that `The 3PAR firmware version check is done at the library level`, then `3PAR OS - 3.2.2 MU6, 3.3.1 MU1` requirement should be replaced by the library name and version.

farhan7500 replied:
Updated the hpe3par_sdk library with the version

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+    secure = module.params["secure"]
+
+    wsapi_url = 'https://%s:8080/api/v1' % storage_system_ip
+    client_obj = client.HPE3ParClient(wsapi_url, secure)
```
`client.HPE3ParClient()` can raise some exceptions which aren't caught:
- [hpe3parclient.exceptions.SSLCertFailed](https://github.com/HewlettPackard/hpe3par_python_sdk/blob/v1.0.1/hpe3par_sdk/client.py#L124)
- [hpe3parclient.exceptions.ConnectionError](https://github.com/HewlettPackard/hpe3par_python_sdk/blob/v1.0.1/hpe3par_sdk/client.py#L130)
- [hpe3parclient.exceptions.UnsupportedVersion](https://github.com/hpe-storage/python-3parclient/blob/master/hpe3parclient/client.py#L216)

farhan7500 replied:
Done

farhan7500 replied:
Done

pilou- commented on lib/ansible/modules/storage/hpe3par/hpe3par_cpg.py:
```diff
@@ -0,0 +1,341 @@
+
+def main():
+
+    module = AnsibleModule(argument_spec=hpe3par.cpg_argument_spec())
```
`check_mode` isn't supported, the documentation should mention that information in `notes` ([see](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block))

farhan7500 replied:
Done

farhan7500 replied:
Added check_mode in the notes

jctanner commented:
cc @dfederlein

farhan7500 commented:
Closing and opening the PR to retrigger CI

dagwieers reviewed (changes requested):
I still have an issue with the unit of each value to be a separate parameter. Basically requiring the user to have to specify 2 parameters for something that should be convenient (both in use as when reviewing playbooks)

farhan7500 commented:
@dagwieers: I have modified the code to now have the capacities as a  single parameter as you suggested. 

Kindly review

farhan7500 commented:
cc @dagwieers : Thanks for the reviews. I have made the changes requested. Please could you review and approve.

Thanks!

farhan7500 commented:
ready_for_review

farhan7500 commented:
cc @dagwieers @pilou- Please could you review the changes I've made after your last review.

Thanks!

farhan7500 commented:
Hi @dagwieers @pilou- 

Please could you review the PR so we can go ahead and submit related PRs for other 3PAR resources. In total we have 11 resources which will be a module each.

I have responded to all the previous review comments and the module has been tested successfully. I have also rebased the code and have passed the Shippable pre checkin tests.

We have passed our internal deadline to close this and it would be really helpful if you could review and approve.

Thanks!

dagwieers commented on lib/ansible/module_utils/hpe3par.py:
```diff
@@ -0,0 +1,84 @@
+# (C) Copyright 2018 Hewlett Packard Enterprise Development LP
```
The syntax for copyright statements is:
```
# Copyright: (c) 2018, Hewlett Packard Enterprise Development LP
```

*(Same for other files below)*

dagwieers commented on lib/ansible/module_utils/hpe3par.py:
```diff
@@ -0,0 +1,84 @@
+# (C) Copyright 2018 Hewlett Packard Enterprise Development LP
+#
+# This program is free software; you can redistribute it and/or modify
```
We have a one-line GPL3 header for sources, however the project only accepts BSD-licensed module_utils files (exceptions exist).
```
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
```

farhan7500 replied:
We will talk to the legal team and get back on this.

farhan7500 replied:
Added one line license headers to each file

dagwieers commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+#!/usr/bin/python
+
+# (C) Copyright 2018 Hewlett Packard Enterprise Development LP
```
The syntax for copyright statements is:
```
# Copyright: (c) 2018, Hewlett Packard Enterprise Development LP
```

dagwieers commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+
+# (C) Copyright 2018 Hewlett Packard Enterprise Development LP
+#
+# This program is free software; you can redistribute it and/or modify
```
The one-line GPL header looks like this:
```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
*(Same for other files below)*

dagwieers commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+    description:
+      - Specifies the name of the domain in which the object will reside.
+  growth_increment:
+    default: -1.0
```
It's weird to have a 'growth_increment' of -1.0 by default.
If this is done to check if the user had specified something, you can simply compare to None.

dagwieers commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+    description:
+      - Specifies whether cerificate need to be installed while communicating
+    type: bool
+    default: false
```
Please use:
```yaml
    default: 'no'
```

dagwieers commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+        set_size: 8
+        high_availability: MAG
+        disk_type: FC
+        secure: false
```
We prefer if examples use 'no' or yes' (more human-readable), so here we prefer:
```yaml
    secure: no
```

*(Also fix the second example)*

dagwieers commented on test/units/modules/storage/hpe3par/test_cpg_3par.py:
```diff
@@ -0,0 +1,261 @@
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of version 3 of the GNU General Public License as
+# published by the Free Software Foundation.  Alternatively, at your
```
Is there a reason here to dual-license (GPL3, Apache 2.0) the unit tests here ?

farhan7500 replied:
Removed the dual license

pilou- commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+    except:
+        module.fail_json(msg="Initializing client failed")
+
+    if storage_system_username is None and storage_system_password is None:
```
Should not that be `or` instead of `and` ?

pilou- commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+    required: true
+  secure:
+    description:
+      - Specifies whether cerificate need to be installed while communicating
```
* s/cerificate/certificate/
* s/installed/validated/

pilou- commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+    except exceptions.SSLCertFailed:
+        module.fail_json(msg="SSL Certificate Failed")
+    except exceptions.ConnectionError:
+        module.fail_json(msg="COnnection Error")
```
s/COnnection/Connection/

pilou- commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,322 @@
+        module.fail_json(msg="COnnection Error")
+    except exceptions.UnsupportedVersion:
+        module.fail_json(msg="Unsupported WSAPI version")
+    except:
```
Bare `except` should be avoided ([see](https://docs.python.org/2/howto/doanddont.html#except)). It might be helpful to include exception in `msg`.

pilou- commented on .github/BOTMETA.yml:
```diff
@@ -306,6 +306,7 @@ files:
   $modules/source_control/github_key.py:
     ignored: erydo
+  $modules/storage/hpe3par/: farhan7500 gautamphegde
   $modules/storage/infinidat/: GR360RY vmalloc
```
Add entries in `BOTMETA.yml` for:
* `$module_utils/hpe3par.py` (or `$module_utils/storage/hpe3par` see my comment below)
* `test/units/modules/storage/hpe3par`
* `lib/ansible/utils/module_docs_fragments/hpe3par`

```
$module_utils/hpe3par.py: &hpe3par
  maintainers: farhan7500 gautamphegde
  support: community
lib/ansible/utils/module_docs_fragments/hpe3par: *hpe3par
test/units/modules/storage/hpe3par: *hpe3par
```

pilou- commented on lib/ansible/module_utils/hpe3par.py:
```diff
@@ -0,0 +1,84 @@
+# (C) Copyright 2018 Hewlett Packard Enterprise Development LP
```
@gundalow told me recently that `module_utils` path should follow the module path, meaning that `lib/ansible/module_utils/storage/hpe3par/__init__.py` should be used. AFAIK, this change is recommended but not mandatory.

gundalow replied:
Would be nice, and it's easier to do now before this gets released

pilou- reviewed (commented):
Sorry for the unresponsiveness :-/ I guess the future resources will be reviewed/merged faster - because you will be able - as maintainers - to approve these changes :)

pilou- commented on .github/BOTMETA.yml:
```diff
@@ -591,6 +592,9 @@ files:
     labels:
       - cloud
     support: community
+  $module_utils/hpe3par.py: &hpe3par
```
It should be `$module_utils/storage/hpe3par.py: &hpe3par`

pilou- replied:
Whoops, I made a mistake a `hpe3par` directory is missing, it should be: `$module_utils/storage/hpe3par/hpe3par.py`.

I created a PR which document `module_utils` tree recommendation: #45561.

pilou- reviewed (commented):
Apart from the `module_utils` path in `BOTMETA.yml`, looks good to me.

pilou- reviewed (approved):
LGTM

farhan7500 commented:
Hi @dagwieers 

I have made the required changes suggested by you. Please could you review.

Thanks!

farhan7500 commented:
Hi @dagwieers 

I have rebased the code, please could you review.

Thanks!

pilou- commented:
@farhan7500 just for you to know, I mentioned this PR and asked for review one week ago at the [ansible core plublic irc meeting](https://meetbot.fedoraproject.org/ansible-meeting/2018-10-09/ansible_core_plublic_irc_meeting.2018-10-09-19.00.log.html#l-121).

pilou- commented on lib/ansible/module_utils/storage/hpe3par/hpe3par.py:
```diff
@@ -0,0 +1,91 @@
+        "set_size": {
+            "required": False,
+            "type": "int",
+            "default": -1
```
It seems `set_size` should be required (then without a default value) since [the default value isn't a valid default value](https://github.com/HewlettPackard/hpe3par_python_sdk/blob/c4e508e6d11ff0fc2d173558d3497a2cb81d0f3b/hpe3par_sdk/client.py#L51).

farhan7500 replied:
As per the [HPE 3PAR Web Sevices API document](https://support.hpe.com/hpsc/doc/public/display?docId=c03606339) - page 84, any negative value for set size is ignored. In that case, if the user does not specify a value for the set size, it is just ignored.

![image](https://user-images.githubusercontent.com/22393330/48303088-ebf73000-e52b-11e8-8a09-8cbc08234fb9.png)

In this case, if the value of set size is not provided, it is defaulted to -1 (a negative value) and is safely ignored.

pilou- replied:
- either when `raid_type` isn't set by user (then `raid_type` value is `None`)
- or when `raid_type` is set by user but `set_size` isn't 

module returns an error and nothing is sent to the HPE 3PAR web service:
https://github.com/ansible/ansible/blob/053ff9998ac44a0ac7392591f1e8261e912351ec/lib/ansible/modules/storage/hpe3par/cpg_3par.py#L130-L135

- either `required_together` must be used ([example](https://github.com/ansible/ansible/blob/053ff9998ac44a0ac7392591f1e8261e912351ec/lib/ansible/modules/crypto/openssl_privatekey.py#L268))
- or `validate_set_size` check should be removed

gundalow replied:
@farhan7500 What are your thoughts on this?

pilou- commented on lib/ansible/modules/storage/hpe3par/cpg_3par.py:
```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/python
+# Copyright: (c) 2018, Hewlett Packard Enterprise Development LP
+# GNU General Public License v3.0+
```
Not sure if the comment must be on the same line than the license name, @gundalow any advice ?

pilou- reviewed (commented):
I checked updates after last review of @dagwieers, all the comments have been taken in account. After reading the [comment about `growth_increment`](https://github.com/ansible/ansible/pull/39656#discussion_r213379093), I added a comment about default value of `set_size`.

@farhan7500 about `set_size`, is a change required ? If so, update the PR, I will check the update, then @gundalow could merge the PR. If not, PR can be merged without further update.

Ideally a user of both Ansible and HPE StoreServ 3PAR device should test the module and report feedback but as far as I know this step isn't required.

gundalow commented on lib/ansible/utils/module_docs_fragments/hpe3par.py:
```diff
@@ -0,0 +1,30 @@
+      required: true
+
+requirements:
+  - Ansible - 2.8
```
This line should be removed, documentation is versioned.

xenlo commented:
I tested this PR and it works for me!

I replaced in our 2 playbooks the calls to `hpe3par_cpg` module (from [HewlettPackard/hpe3par_ansible_module](https://github.com/HewlettPackard/hpe3par_ansible_module)) the by this fresh `cpg_3par` ansible core module and I get the expected result!

**Testing tasks**
from `prepare-hpe3par.yml` playbook:
```
[…]
    - name: Create Common Provisioning Group
      cpg_3par:
        storage_system_ip: "{{ hpe3par_ip }}"
        storage_system_username: "{{ hpe3par_user }}"
        storage_system_password: "{{ hpe3par_password }}"
        state: present
        cpg_name: "{{ cpg_name }}"
        raid_type: R5
        set_size: 8
        high_availability: MAG
        disk_type: FC
[…]
```

from `decomm-hpe3par.yml` playbook:
```
[…]
    - name: Delete all Common Provisioning Groups
      cpg_3par:
        storage_system_ip: "{{ hpe3par_ip }}"
        storage_system_username: "{{ hpe3par_user }}"
        storage_system_password: "{{ hpe3par_password }}"
        state: absent
        cpg_name: "{{ item }}"
      loop: "{{ cpgs.json.members | map(attribute='name') | list }}"
[…]
```


**Ansible version**
```
$ ansible --version
ansible 2.8.0.dev0 (testing_PR39656 af5d84b70b) last updated 2018/12/13 11:30:10 (GMT +200)
  config file = $HOME/Projects/97/etc/ansible.cfg
  configured module search path = [u'$HOME/.virtualenvs/97/local/lib/python2.7/site-packages/ara/plugins/modules', u'$HOME/Projects/97/library']
  ansible python module location = $HOME/Projects/test_cpg_3par/ansible/lib/ansible
  executable location = $HOME/Projects/test_cpg_3par/ansible/bin/ansible
  python version = 2.7.13 (default, Sep 26 2018, 18:42:22) [GCC 6.3.0 20170516]
```

xenlo commented:
Thanks @farhan7500,
Looking forward for the merge of next modules!

xenlo reviewed (approved):
Successfully tested this module.

more details: https://github.com/ansible/ansible/pull/39656#issuecomment-446932561

dagwieers commented:
> But the fact that companies names can change is not the right reason. Sometimes, vendors also change name of their products… 

Sure, so you want to increase the probability that we have to rename in the future by also including the vendor name?

xenlo commented:
> 
> 
> > But the fact that companies names can change is not the right reason. Sometimes, vendors also change name of their products…
> 
> Sure, so you want to increase the probability that we have to rename in the future by also including the vendor name?

No not at all. I was actually thinking out loud, trying to find something the most persistent possible name that will not start with number. I found `yellowSAN_*` which is not good match 😄. So in the end I would go for `3par_*`, too.

thedoubl3j reviewed (changes requested):
@farhan7500 please back out the previous commit `de60766cd5f6861f0cd1577ee474f9c56aa111b3`. 

For every **new** module, each module and its descriptive unit tests will need to be contained in its own PR.

thedoubl3j reviewed (commented):
might have to squash commits but other than that, my request was made in error. 

will ping @gundalow when he gets back online

samccann commented on lib/ansible/modules/storage/hpe3par/ss_3par_cpg.py:
```diff
@@ -0,0 +1,289 @@
+      - Specifies the name of the domain in which the object will reside.
+  growth_increment:
+    description:
+      - Specifies the growth increment the amount of logical disk storage
```
Would be good to mention if this is measured in MiB (based on test code that seems correct?) same for the other growth_* options.

samccann commented on lib/ansible/modules/storage/hpe3par/ss_3par_cpg.py:
```diff
@@ -0,0 +1,289 @@
+    required: true
+  secure:
+    description:
+      - Specifies whether certificate need to be validated while communicating
```
```suggestion
      - Specifies whether the certificate needs to be validated while communicating.
```

samccann commented on lib/ansible/modules/storage/hpe3par/ss_3par_cpg.py:
```diff
@@ -0,0 +1,289 @@
+        module.fail_json(msg='the python hpe3par_sdk library is required (https://pypi.org/project/hpe3par_sdk)')
+
+    if len(module.params["cpg_name"]) < 1 or len(module.params["cpg_name"]) > 31:
+        module.fail_json(msg="CPG name must be atleast 1 character and not more than 31 characters")
```
```suggestion
        module.fail_json(msg="CPG name must be at least 1 character and not more than 31 characters")
```

samccann reviewed (commented):
The doc portion looks good in general. I've added a few comments/suggestions that would improve it.  Sorry for the lateness on this docs review!

thedoubl3j reviewed (approved):
ran tests locally and succeeded. everything looks good to me

farhan7500 committed:
Modified the documentation based on the review comments

farhan7500 commented:
@thedoubl3j I have squashed the commits but ansibot is still asking me to rebase. What do I need to do to resolve this? Thanks!

samccann commented:
You may find this helpful for rebasing:
https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html?highlight=rebase

farhan7500 committed:
Fixed the pep8 sanity issue

farhan7500 committed:
Moved the doc fragment file based on latest changes

jctanner commented:
!waffling needs_revision

farhan7500 committed:
Reverting inadvertent rebase changes

thedoubl3j commented:
tests pass, CI is passing. Ping @gundalow

dagwieers reviewed (approved):
So I still have many concerns:
- This module is doing its own unit-size calculation, we should really have a single interface for this so all modules do this uniformly
- The name is still hpe3par, which I think is more problematic than it should be
- Parameters lack a type in the documentation

But I don't want to drag this any longer.

gundalow commented on lib/ansible/module_utils/storage/hpe3par/hpe3par.py:
```diff
@@ -0,0 +1,90 @@
+from ansible.module_utils import basic
+
+
+def convert_to_binary_multiple(size_with_unit):
```
Would `human_to_bytes` from `module_utils/basic.py` be of use instead?

farhan7500 replied:
We have changed the method to use `human_to_bytes` internally in this method.

`size_kib = basic.human_to_bytes(size)`

The only addition here is that we are changing the user provided units (MiB, KiB, GiB) into M, T and K that human_to_bytes method expects.

farhan7500 committed:
Fixed inadvertent change during rebase

gundalow commented on lib/ansible/module_utils/storage/hpe3par/hpe3par.py:
```diff
@@ -0,0 +1,90 @@
+    return int(size_kib / (1024 * 1024))
+
+
+storage_system_spec = {
```
Appologies I missed this when reviewing. The common arguments must be moved to `module_utils/storage/hpe3par/`.

dagwieers replied:
That's already the case, no ?
![screenshot from 2019-02-26 15-41-20](https://user-images.githubusercontent.com/388198/53420933-0242d200-39dd-11e9-947a-bd6d6619a599.png)

gundalow replied:
ah, looks like you are right https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/storage/hpe3par/hpe3par.py#L26-L45

@farhan7500 In a follow up PR can you please remove the duplicated options.

gundalow commented:
Adding `type` and string for `version_added` done in https://github.com/ansible/ansible/pull/52992
